File: ios/Classes/AesCtrCryptography.swift
//
//  AesCtrCryptography.swift
//  aes_ctr
//
//  Created by Endtry on 9/10/2563 BE.
//

import Foundation
import CryptoSwift

/**
 * https://flutter.dev/docs/development/packages-and-plugins/developing-packages
 */
class AesCtrCryptography: Cryptography {
    
    func encrypt(counter: NSNumber, secretKey: FlutterStandardTypedData, plaintext: String) -> String {
        do {
            let key = [UInt8](secretKey.data)
            var iv = [UInt8](repeating: 0x0, count: 16)
            iv[15] = counter.uint8Value
            let aes = try AES(key: key, blockMode: CTR(iv: iv), padding: .noPadding)
            let encrypted = try aes.encrypt(plaintext.bytes)
            let cipherText = encrypted.toHexString()
            return cipherText
        } catch {
            return "Encrypt error"
        }
    }
    
    func decrypt(counter: NSNumber, secretKey: FlutterStandardTypedData, cipher: String) -> String {
        do {
            let key = [UInt8](secretKey.data)
            var iv = [UInt8](repeating: 0x0, count: 16)
            iv[15] = counter.uint8Value
            let hexArray = Array<UInt8>.init(hex: cipher)
            let aes = try AES(key: key, blockMode: CTR(iv: iv), padding: .noPadding)
            let decrypted = try aes.decrypt(hexArray)
            if let decryptedTxt = String(bytes: Data(decrypted).bytes, encoding: .utf8) {
                return decryptedTxt
            }
            return ""
        } catch {
            return "Decrypt error"
        }
    }
    
    static func getInstance() -> Cryptography {
        return AesCtrCryptography()
    }
    
}

File: android/src/main/kotlin/com/prongbang/aesctr/crypto/AesCtrCryptography.kt
package com.prongbang.aesctr.crypto

import android.annotation.SuppressLint
import org.bouncycastle.util.Arrays
import java.nio.charset.StandardCharsets
import java.security.Security
import javax.crypto.Cipher
import javax.crypto.spec.IvParameterSpec
import javax.crypto.spec.SecretKeySpec

/**
 * https://flutter.dev/docs/development/packages-and-plugins/developing-packages
 */
class AesCtrCryptography : Cryptography {

	init {
		Security.addProvider(org.bouncycastle.jce.provider.BouncyCastleProvider())
	}

	@SuppressLint("GetInstance")
	override fun encrypt(counter: Int, secretKey: ByteArray, plaintext: String): String {
		val iv = ByteArray(16)
		Arrays.fill(iv, 0.toByte())
		iv[15] = counter.toByte()

		val key = SecretKeySpec(secretKey, "AES")
		val cipher = Cipher.getInstance("AES/CTR/NoPadding", "BC")

		// Encrypt
		val ivParam = IvParameterSpec(iv)
		cipher.init(Cipher.ENCRYPT_MODE, key, ivParam)
		val encrypted = cipher.doFinal(plaintext.toByteArray())

		return CryptoHelper.toHex(encrypted) ?: ""
	}

	@SuppressLint("GetInstance")
	override fun decrypt(counter: Int, secretKey: ByteArray, cipher: String): String {
		val iv = ByteArray(16)
		Arrays.fill(iv, 0.toByte())
		iv[15] = counter.toByte()

		val key = SecretKeySpec(secretKey, "AES")
		val cipherInstance = Cipher.getInstance("AES/CTR/NoPadding", "BC")

		val ivParam = IvParameterSpec(iv)
		cipherInstance.init(Cipher.ENCRYPT_MODE, key, ivParam)

		val param = cipherInstance.parameters

		val encrypted = CryptoHelper.toByte(cipher)

		cipherInstance.init(Cipher.DECRYPT_MODE, key, param)
		val decryptedByte = cipherInstance.doFinal(encrypted)

		return String(decryptedByte, StandardCharsets.UTF_8)
	}

	companion object {
		fun getInstance(): Cryptography = AesCtrCryptography()
	}
}